File: tools/util_test.py
# Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
import os

from test_util import DenoTestCase, run_tests
from util import (parse_exit_code, shell_quote_win, root_path)


class TestUtil(DenoTestCase):
    def test_parse_exit_code(self):
        assert parse_exit_code('hello_error54_world') == 54
        assert parse_exit_code('hello_error_world') == 1
        assert parse_exit_code('hello_world') == 0

    def test_shell_quote_win(self):
        assert shell_quote_win('simple') == 'simple'
        assert shell_quote_win(
            'roof/\\isoprojection') == 'roof/\\isoprojection'
        assert shell_quote_win('with space') == '"with space"'
        assert shell_quote_win('embedded"quote') == '"embedded""quote"'
        assert shell_quote_win(
            'a"b""c\\d\\"e\\\\') == '"a""b""""c\\d\\\\""e\\\\\\\\"'

    def test_executable_exists(self):
        assert os.path.exists(self.deno_exe)


if __name__ == '__main__':
    run_tests()

File: tools/util_test.ts
// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
import * as deno from "deno";
import { assert, testPerm, assertEqual } from "../js/test_util.ts";
import { findFiles } from "./util.ts";

const testDir = "tools/testdata/find_files_testdata";

// Sorts and replace backslashes with slashes.
const normalize = files => files.map(f => f.replace(/\\/g, "/")).sort();

testPerm({ read: true }, function testFindFiles() {
  const files = findFiles([testDir], [".ts", ".md"]);
  assertEqual(normalize(files), [
    `${testDir}/bar.md`,
    `${testDir}/bar.ts`,
    `${testDir}/foo.md`,
    `${testDir}/foo.ts`,
    `${testDir}/subdir0/bar.ts`,
    `${testDir}/subdir0/foo.ts`,
    `${testDir}/subdir0/subdir0/bar.ts`,
    `${testDir}/subdir0/subdir0/foo.ts`,
    `${testDir}/subdir1/bar.ts`,
    `${testDir}/subdir1/foo.ts`
  ]);
});

testPerm({ read: true }, function testFindFilesDepth() {
  const files = findFiles([testDir], [".ts", ".md"], { depth: 1 });
  assertEqual(normalize(files), [
    `${testDir}/bar.md`,
    `${testDir}/bar.ts`,
    `${testDir}/foo.md`,
    `${testDir}/foo.ts`
  ]);
});

testPerm({ read: true }, function testFindFilesSkip() {
  const files = findFiles([testDir], [".ts", ".md"], {
    skip: ["foo.md", "subdir1"]
  });
  assertEqual(normalize(files), [
    `${testDir}/bar.md`,
    `${testDir}/bar.ts`,
    `${testDir}/foo.ts`,
    `${testDir}/subdir0/bar.ts`,
    `${testDir}/subdir0/foo.ts`,
    `${testDir}/subdir0/subdir0/bar.ts`,
    `${testDir}/subdir0/subdir0/foo.ts`
  ]);
});

testPerm({ read: false }, function testFindFilesPerm() {
  let caughtError = false;
  try {
    const files = findFiles([testDir], [".ts", ".md"]);
  } catch (e) {
    caughtError = true;
    assertEqual(e.kind, deno.ErrorKind.PermissionDenied);
    assertEqual(e.name, "PermissionDenied");
  }
  assert(caughtError);
});
